File: riveredge-core/src/models/tenant.py
"""
组织模型模块

定义组织数据模型，用于多组织系统的组织管理
"""

from enum import Enum
from typing import Optional, Dict, Any

from tortoise import fields

from models.base import BaseModel


class TenantStatus(str, Enum):
    """
    组织状态枚举
    
    定义组织的可用状态
    """
    ACTIVE = "active"          # 激活状态
    INACTIVE = "inactive"      # 未激活状态
    EXPIRED = "expired"        # 已过期状态
    SUSPENDED = "suspended"    # 已暂停状态


class TenantPlan(str, Enum):
    """
    组织套餐枚举
    
    定义组织的套餐类型
    """
    TRIAL = "trial"                    # 体验套餐
    BASIC = "basic"                    # 基础套餐
    PROFESSIONAL = "professional"     # 专业套餐
    ENTERPRISE = "enterprise"          # 企业套餐


class Tenant(BaseModel):
    """
    组织模型

    用于管理 SaaS 多组织系统中的组织信息。
    组织表本身不包含 tenant_id（因为组织本身就是组织的定义），
    但继承 BaseModel 以保持一致性（tenant_id 为 None）。

    Attributes:
        id: 组织 ID（主键）
        name: 组织名称
        domain: 组织域名（用于子域名访问）
        status: 组织状态（active, inactive, expired, suspended）
        plan: 组织套餐（basic, professional, enterprise）
        settings: 组织配置（JSONB 存储）
        max_users: 最大用户数限制
        max_storage: 最大存储空间限制（MB）
        expires_at: 过期时间（可选）
        created_at: 创建时间
        updated_at: 更新时间
    """

    id = fields.IntField(primary_key=True, description="组织 ID（主键）")
    name = fields.CharField(max_length=100, description="组织名称")
    domain = fields.CharField(max_length=100, unique=True, description="组织域名（用于子域名访问）")
    status = fields.CharEnumField(
        enum_type=TenantStatus,
        default=TenantStatus.INACTIVE,
        description="组织状态"
    )
    plan = fields.CharEnumField(
        enum_type=TenantPlan,
        default=TenantPlan.BASIC,
        description="组织套餐"
    )
    settings = fields.JSONField(default=dict, description="组织配置（JSONB 存储）")
    max_users = fields.IntField(default=10, description="最大用户数限制")
    max_storage = fields.IntField(default=1024, description="最大存储空间限制（MB）")
    expires_at = fields.DatetimeField(null=True, description="过期时间（可选）")

    class Meta:
        """
        模型元数据
        """
        table = "tree_tenants"  # 表名必须包含模块前缀（tree_ - 租户管理）
        indexes = [
            ("domain",),  # 域名索引
            ("status",),  # 状态索引
            ("plan",),    # 套餐索引
        ]
    
    def __str__(self) -> str:
        """
        返回组织的字符串表示
        
        Returns:
            str: 组织字符串表示
        """
        return f"<Tenant(id={self.id}, name={self.name}, domain={self.domain})>"
    
    async def is_active(self) -> bool:
        """
        检查组织是否处于激活状态
        
        Returns:
            bool: 如果组织状态为 active 则返回 True，否则返回 False
        """
        return self.status == TenantStatus.ACTIVE
    
    async def is_expired(self) -> bool:
        """
        检查组织是否已过期
        
        Returns:
            bool: 如果组织已过期则返回 True，否则返回 False
        """
        if self.expires_at is None:
            return False
        from datetime import datetime
        return datetime.now() > self.expires_at

File: riveredge-stem/src/services/tenant.ts
/**
 * 组织 API 服务
 * 
 * 提供组织管理相关的 API 接口
 * 注意：组织管理通常需要超级管理员权限
 */

// 使用 apiRequest 统一处理 HTTP 请求


import { apiRequest } from './api';

/**
 * 组织状态枚举
 */
export enum TenantStatus {
  ACTIVE = 'active',
  INACTIVE = 'inactive',
  EXPIRED = 'expired',
  SUSPENDED = 'suspended',
}

/**
 * 组织套餐枚举
 */
export enum TenantPlan {
  TRIAL = 'trial',           // 体验套餐
  BASIC = 'basic',           // 基础版
  PROFESSIONAL = 'professional',  // 专业版
  ENTERPRISE = 'enterprise',  // 企业版
}

/**
 * 套餐配置接口
 */
export interface PackageConfig {
  name: string;
  max_users: number;
  max_storage_mb: number;
  allow_pro_apps: boolean;
  description: string;
}

/**
 * 所有套餐配置
 */
export interface AllPackageConfigs {
  [key: string]: PackageConfig;
}

/**
 * 组织信息接口
 */
export interface Tenant {
  id: number;
  name: string;
  domain: string;
  status: TenantStatus;
  plan: TenantPlan;
  settings: Record<string, any>;
  max_users: number;
  max_storage: number;
  expires_at?: string;
  created_at: string;
  updated_at: string;
}

/**
 * 组织列表查询参数
 */
export interface TenantListParams {
  page?: number;
  page_size?: number;
  status?: TenantStatus;
  plan?: TenantPlan;
  keyword?: string;  // 关键词搜索（组织名称、域名，使用 OR 逻辑）
  name?: string;  // 组织名称搜索（精确搜索）
  domain?: string;  // 域名搜索（精确搜索）
  sort?: string;  // 排序字段（如 'name', 'created_at'）
  order?: 'asc' | 'desc';  // 排序方向
}

/**
 * 组织列表响应数据
 */
export interface TenantListResponse {
  items: Tenant[];
  total: number;
  page: number;
  page_size: number;
}

/**
 * 创建组织数据
 */
export interface CreateTenantData {
  name: string;
  domain: string;
  status?: TenantStatus;
  plan?: TenantPlan;
  settings?: Record<string, any>;
  max_users?: number;
  max_storage?: number;
  expires_at?: string;
}

/**
 * 更新组织数据
 */
export interface UpdateTenantData {
  name?: string;
  domain?: string;
  status?: TenantStatus;
  plan?: TenantPlan;
  settings?: Record<string, any>;
  max_users?: number;
  max_storage?: number;
  expires_at?: string;
}

/**
 * 获取组织列表
 * 
 * @param params - 查询参数
 * @param isSuperAdmin - 是否为超级管理员接口（默认 false，使用 /tenants；true 时使用 /superadmin/tenants）
 * @returns 组织列表响应数据
 */
export async function getTenantList(
  params: TenantListParams,
  isSuperAdmin: boolean = false
): Promise<TenantListResponse> {
  const endpoint = isSuperAdmin ? '/superadmin/tenants' : '/tenants';
  return apiRequest<TenantListResponse>(endpoint, {
    method: 'GET',
    params,
  });
}

/**
 * 获取组织详情
 * 
 * @param tenantId - 组织 ID
 * @returns 组织详情
 */
export async function getTenantById(tenantId: number): Promise<Tenant> {
  return apiRequest<Tenant>(`/tenants/${tenantId}`, {
    method: 'GET',
  });
}

/**
 * 创建组织
 * 
 * @param data - 组织创建数据
 * @returns 创建的组织
 */
export async function createTenant(data: CreateTenantData): Promise<Tenant> {
  return apiRequest<Tenant>('/tenants', {
    method: 'POST',
    data,
  });
}

/**
 * 更新组织
 * 
 * @param tenantId - 组织 ID
 * @param data - 组织更新数据
 * @returns 更新后的组织
 */
export async function updateTenant(tenantId: number, data: UpdateTenantData): Promise<Tenant> {
  return apiRequest<Tenant>(`/tenants/${tenantId}`, {
    method: 'PUT',
    data,
  });
}

/**
 * 删除组织（软删除）
 * 
 * @param tenantId - 组织 ID
 */
export async function deleteTenant(tenantId: number): Promise<void> {
  return apiRequest<void>(`/tenants/${tenantId}`, {
    method: 'DELETE',
  });
}

/**
 * 激活组织
 * 
 * @param tenantId - 组织 ID
 * @returns 更新后的组织
 */
export async function activateTenant(tenantId: number): Promise<Tenant> {
  return apiRequest<Tenant>(`/tenants/${tenantId}/activate`, {
    method: 'POST',
  });
}

/**
 * 停用组织
 * 
 * @param tenantId - 组织 ID
 * @returns 更新后的组织
 */
export async function deactivateTenant(tenantId: number): Promise<Tenant> {
  return apiRequest<Tenant>(`/tenants/${tenantId}/deactivate`, {
    method: 'POST',
  });
}

/**
 * 获取所有套餐配置
 * 
 * @returns 所有套餐配置
 */
export async function getPackageConfigs(): Promise<AllPackageConfigs> {
  return apiRequest<AllPackageConfigs>('/tenants/packages/config', {
    method: 'GET',
  });
}

/**
 * 获取指定套餐配置
 * 
 * @param plan - 套餐类型
 * @returns 套餐配置
 */
export async function getPackageConfig(plan: TenantPlan): Promise<PackageConfig> {
  return apiRequest<PackageConfig>(`/tenants/packages/${plan}/config`, {
    method: 'GET',
  });
}

/**
 * 组织使用量统计接口
 */
export interface TenantUsage {
  tenant_id: number;
  user_count: number;
  max_users: number;
  storage_used_mb: number;
  max_storage_mb: number;
  user_usage_percent: number;
  storage_usage_percent: number;
  warnings: string[];
}

/**
 * 获取组织使用量统计
 * 
 * @param tenantId - 组织 ID
 * @returns 组织使用量统计
 */
export async function getTenantUsage(tenantId: number): Promise<TenantUsage> {
  return apiRequest<TenantUsage>(`/tenants/${tenantId}/usage`, {
    method: 'GET',
  });
}

/**
 * 组织活动日志接口
 */
export interface TenantActivityLog {
  id: number;
  tenant_id: number;
  action: string;
  description: string;
  operator_id?: number;
  operator_name?: string;
  created_at: string;
}

/**
 * 组织活动日志列表查询参数
 */
export interface TenantActivityLogListParams {
  page?: number;
  page_size?: number;
  action?: string;
}

/**
 * 组织活动日志列表响应数据
 */
export interface TenantActivityLogListResponse {
  items: TenantActivityLog[];
  total: number;
  page: number;
  page_size: number;
}

/**
 * 获取组织活动日志列表
 * 
 * @param tenantId - 组织 ID
 * @param params - 查询参数
 * @returns 组织活动日志列表响应数据
 */
export async function getTenantActivityLogs(
  tenantId: number,
  params: TenantActivityLogListParams = {}
): Promise<TenantActivityLogListResponse> {
  return apiRequest<TenantActivityLogListResponse>(`/tenants/${tenantId}/activity-logs`, {
    method: 'GET',
    params,
  });
}

